### ISSUE TITLE
Errors are not caught in the sockets hook

### ISSUE THREAD

**Issue by andriussev** [balderdashy/sails#2968]
I've been having trouble using sockets as I often get a nasty error that is not caught and crashes the app:

> /apps/sails/node_modules/sails/node_modules/sails-hook-sockets/lib/sails.sockets/get-socket-by-id.js:45
throw ERRORPACK.NO_SUCH_SOCKET('Cannot find socket with id=`%s`', id);
^
Error (SAILS:HOOK:SOCKETS:NO_SUCH_SOCKET):: Cannot find socket with id=`0H11WvjCCxfUydlKAAG8`
  at new constructor (/apps/sails_app/node_modules/sails/node_modules/sails-hook-sockets/standalone/create-error-constructor.js:38:16)
  at Errorpack.factory [as NO_SUCH_SOCKET] (/apps/sails_app/node_modules/sails/node_modules/sails-hook-sockets/standalone/create-error-factory.js:34:12)
  at Object.getSocketById [as get] (/apps/sails_app/node_modules/sails/node_modules/sails-hook-sockets/lib/sails.sockets/get-socket-by-id.js:45:23)
  at /apps/sails_app/node_modules/sails/node_modules/sails-hook-sockets/lib/sails.sockets/emit-to-sockets.js:47:19
  at Function.forEach (/apps/sails_app/node_modules/sails/node_modules/lodash/dist/lodash.js:3298:15)
  at Object.emitToSockets [as emit] (/apps/sails_app/node_modules/sails/node_modules/sails-hook-sockets/lib/sails.sockets/emit-to-sockets.js:46:7)
  at /apps/sails_app/api/controllers/PollController.js:217:31
  at /apps/sails_app/node_modules/sails/node_modules/async/lib/async.js:451:17
  at /apps/sails_app/node_modules/sails/node_modules/async/lib/async.js:441:17
  at _each (/apps/sails_app/node_modules/sails/node_modules/async/lib/async.js:46:13)
  at Object.taskComplete (/apps/sails_app/node_modules/sails/node_modules/async/lib/async.js:440:13)
  at processImmediate [as _immediateCallback] (timers.js:354:15)

The error does happen if I run the app in production mode too.
I think, I only got the error once when testing it locally but running it live with 300+ active users makes the app crash almost once every 10 minutes.

Is there at least a nasty work-around I could quickly implement?

**Comment by nonpc** [comment]
I have done a quick solution to this. Though, it really needs a more serious one.

*/node_modules/sails/node_modules/sails-hook-sockets/lib/sails.sockets/emit-to-sockets.js*, change line with line (at the bottom):
>     _.each(socketIDs, function(socketID) {
      sails.sockets.get(socketID).emit(eventName, data);
    });

To

>     _.each(socketIDs, function(socketID) {
        var socket_to_emit_to = sails.sockets.get(socketID);
        if(socket_to_emit_to != null) {
            sails.sockets.get(socketID).emit(eventName, data);
        }
    });

Edit */node_modules/sails/node_modules/sails-hook-sockets/lib/sails.sockets/get-socket-by-id.js*,change line with line (also at the bottom):
>     var foundSocket = namespace.connected[id+''];
    if (!foundSocket) {
      throw ERRORPACK.NO_SUCH_SOCKET('Cannot find socket with id=`%s`', id);
    }

To

>     var foundSocket = namespace.connected[id+''];
    if (!foundSocket) {
        return null;
      throw ERRORPACK.NO_SUCH_SOCKET('Cannot find socket with id=`%s`', id);
    }

That does mean I won't be throwing any more errors. So, if you're getting this or any of the other ones in the function (like '**`sails.sockets.get()` cannot lookup socket w/o an id**' or '**Cannot lookup socket w/ invalid id**' or '**Cannot find namespace with pathname identifier**'), you can do the return. Though, it should be properly fixed so you can actually see the errors happening.

**Comment by sgress454** [comment]
Thanks for the report @nonpc.  Can you please also post a link to this in the [sails-hook-sockets repo](https://github.com/balderdashy/sails-hook-sockets)?

**Comment by sailsbot** [comment]
Thanks for posting, @andriussev.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
